ArtOfIllusion/src/artofillusion/CreatePolygonTool.java

```diff
@@ -66,155 +71,220 @@ public String getToolTipText()
+    if (drawFilled)
+    {
+      vertex = new Vec3 [sides+1];
+      vertex[vertex.length-1] = new Vec3();
+    }
+    else
+      vertex = new Vec3 [sides];
+    for (int i = 0; i < sides; i++)
+      vertex[i] = new Vec3(sine[i], cosine[i], 0.0);
+
+    Object3D object;
+    if (drawFilled)
+    {
+      int faces[][] = new int [sides][];
+      faces[0] = new int [] {sides-1, 0, sides};
+      for (int i = 1; i < sides; i++)
+        faces[i] = new int [] {i-1, i, sides};
+      object = new TriangleMesh(vertex, faces);
+      ((TriangleMesh) object).setSmoothingMethod(shape);
+    }
+    else
+    {
+      float[] s = new float [sides];
+      for (int i = 0; i < vertex.length; i++)
+        s[i] = 1.0f;
+      object = new Curve(vertex, s, shape, true);
+    }
+    return object;
+  }
+
+  private void update3D(WidgetMouseEvent e, ViewerCanvas view)
   {
     Camera cam = view.getCamera();
     Point dragPoint = e.getPoint();
-    Vec3 vertex[], orig, ydir, zdir, temp;
-    CoordinateSystem coords;
-    double scale;
-    float s[] = null;
-    Object3D obj;
-
-    findPoints(dragPoint, e.isShiftDown());
-    if (e.isControlDown())
+    ydir.set(cam.getCameraCoordinates().getUpDirection());
+    zdir.set(cam.getCameraCoordinates().getZDirection());
+    zdir.scale(-1.0);
+    xdir = ydir.cross(zdir);
+
+    double w, h, xscale, yscale;
+
+    Vec3 click3D = cam.convertScreenToWorld(clickPoint, view.getDistToPlane());
+    Vec3 drag3D = cam.convertScreenToWorld(dragPoint, view.getDistToPlane());
+    w = drag3D.minus(click3D).dot(xdir);
+    h = drag3D.minus(click3D).dot(ydir);
+
+    // Update the size of the polygon
+
+    if (centered)
     {
-      vertex = new Vec3 [points.length+1];
-      vertex[points.length] = new Vec3();
+      xscale = Math.abs(w/(maxsine-minsine)*2.0);
+      if (h > 0)
+        yscale = Math.abs(h/(mincosine));
+      else
+        yscale = Math.abs(h/(maxcosine));
     }
     else
     {
-      vertex = new Vec3 [points.length];
-      s = new float [points.length];
-      for (int i = 0; i < points.length; i++)
-        s[i] = 1.0f;
+      xscale = Math.abs(w/(maxsine-minsine));
+      yscale = Math.abs(h/(maxcosine-mincosine));
     }
-    orig = cam.convertScreenToWorld(new Point((int) centerx, (int) centery), view.getDistToPlane(), false);
-    temp = cam.convertScreenToWorld(new Point(1+(int) centerx, (int) centery), view.getDistToPlane(), false);
-    scale = temp.minus(orig).length();
-    for (int i = 0; i < points.length; i++)
-      vertex[i] = new Vec3(scale*(points[i].x-centerx), -scale*(points[i].y-centery), 0.0);
+    if (equilateral)
+      xscale = yscale = Math.max(xscale, yscale);
 
-    // Find the object's coordinate system.
+    Vec3[] vertex = ((Mesh)(objInfo.object)).getVertexPositions();
+    for (int i = 0; i < sides; i++)
+    {
+      vertex[i].x = sine[i]*xscale;
+      vertex[i].y = -cosine[i]*yscale;
+    }
+    ((Mesh)(objInfo.object)).setVertexPositions(vertex);
```

Comment:
And another here. `objectInfo.object` does not need to be wrapped.

Reply:
Thanks.